### PR TITLE
 fix for memcpy-param-overlap in the gr-qtgui/waterfall_sink

### PR DESCRIPTION
**Pull request by hcab14**
For odd FFT sizes there is a memcpy overlap in gr-qtgui/waterfall_sink:

```
==16811==ERROR: AddressSanitizer: memcpy-param-overlap: memory ranges [0x6210003de100,0x6210003de900) and [0x6210003de8fc, 0x6210003df0fc) overlap
    #0 0x7fcfd5b2c425  (/usr//lib/gcc/x86_64-linux-gnu/7/libasan.so+0x79425)
    #1 0x7fcfc48e2d69 in memcpy /usr/include/x86_64-linux-gnu/bits/string_fortified.h:34
    #2 0x7fcfc48e2d69 in gr::qtgui::waterfall_sink_c_impl::fft(float*, std::complex<float> const*, int) /home/hcab14/gnuradio/gr-qtgui/lib/waterfall_sink_c_impl.cc:412
    #3 0x7fcfc48e542f in gr::qtgui::waterfall_sink_c_impl::work(int, std::vector<void const*, std::allocator<void const*> >&, std::vector<void*, std::allocator<void*> >&) /home/hcab14/gnuradio/gr-qtgui/lib/waterfall_sink_c_impl.cc:566
```

The memcpy overlap happens during the cyclic shift of the FFT bins after the FFT.

A related problem (heap buffer overflow) happens at the analogous place in the qtgui freq_sink, and the 2nd commit fixes https://github.com/gnuradio/gnuradio/issues/2407

Note that in parallel this issue is being worked on here https://github.com/gnuradio/gnuradio/pull/2486

### CODE REVIEW

**Comment by mormj** [comment]
I like the fft shift object (need to do the same thing as you mentioned in #2486 - should it be more generally available under gr-fft?

File: gr-qtgui/lib/fft_shift.h [hunk 1]
```diff
@@ -0,0 +1,67 @@
+/* -*- c++ -*- */
+/*
+ * Copyright 2012,2014-2015 Free Software Foundation, Inc.
```
Comment (mbr0wn) [inline gr-qtgui/lib/fft_shift.h]: 2019

File: gr-qtgui/lib/fft_shift.h [hunk 2]
```diff
@@ -0,0 +1,67 @@
+/* -*- c++ -*- */
```
Comment (mbr0wn) [inline gr-qtgui/lib/fft_shift.h]: Can we move this into `gr-fft` please?

File: gr-qtgui/lib/fft_shift.h [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ */
+
+#ifndef INCLUDED_QTGUI_FFT_SHIFT_H
+#define INCLUDED_QTGUI_FFT_SHIFT_H
+
+#include <algorithm>
+#include <vector>
+
+namespace gr {
+  namespace qtgui {
```
Comment (mbr0wn) [inline gr-qtgui/lib/fft_shift.h]:
```suggestion
  namespace fft {
```

File: gr-qtgui/lib/fft_shift.h [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    class fft_shift {
+    public:
+      fft_shift(size_t fft_length)
+	: d_fftlen()
```
Comment (mbr0wn) [inline gr-qtgui/lib/fft_shift.h]: If you want to default-initialize these, I would recommend doing it downstairs instead.

File: gr-qtgui/lib/fft_shift.h [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      }
+      void shift(float* data, size_t fft_len) {
+	resize(fft_len);
+	std::copy_n(data,          d_lenpos, d_buf.begin());
+	std::copy_n(data+d_lenpos, d_lenneg, data);
+	std::copy_n(d_buf.begin(), d_lenpos, data+d_lenneg);
+      }
+    protected:
+    private:
+      size_t d_fftlen;
```
Comment (mbr0wn) [inline gr-qtgui/lib/fft_shift.h]:
```suggestion
      size_t d_fftlen = 0;
```

Comment (mbr0wn) [reply gr-qtgui/lib/fft_shift.h]: (and so on)

File: gr-qtgui/lib/fft_shift.h (hunk 1 above)
Comment (hcab14) [reply gr-qtgui/lib/fft_shift.h]: I have added "2019" also to changed files in gr-qtgui. Is this ok?

File: gr-qtgui/lib/fft_shift.h (hunk 2 above)
Comment (hcab14) [reply gr-qtgui/lib/fft_shift.h]: Done

File: gr-qtgui/lib/fft_shift.h (hunk 3 above)
Comment (hcab14) [reply gr-qtgui/lib/fft_shift.h]: ok

File: gr-qtgui/lib/fft_shift.h (hunk 4 above)
Comment (hcab14) [reply gr-qtgui/lib/fft_shift.h]: I realized there  is no need to default initialize and then to set the correct values. It is now done in the constructor in one go (not calling `resize`).

**Comment by hcab14** [comment]
Hi @mbr0wn 

thank you for your review. 

`fft_shift` now lives in gr-fft and became a template, as is might be useful to shift FFT bins and not only |FFT bin|.

**Comment by mbr0wn** [comment]
Merged manually (squashed last 3  commits).

**Pull request closed without merging**